Vostok.Commons.Collections/ConcurrentBoundedQueue.cs

```diff
@@ -127,29 +138,52 @@ public int Drain(T[] buffer, int index, int count)
         /// <summary>
         /// Asynchronously waits until something is available to <see cref="Drain"/>.
         /// </summary>
-        public Task WaitForNewItemsAsync() => canDrain.Task;
+        public Task WaitForNewItemsAsync() => canDrainAny.Task;
 
         /// <summary>
         /// Asynchronously waits until something is available to <see cref="Drain"/> or the provided <paramref name="timeout"/> expires.
         /// </summary>
         /// <returns><c>true</c> if there is something to drain, <c>false</c> otherwise.</returns>
         public async Task<bool> TryWaitForNewItemsAsync(TimeSpan timeout)
         {
-            if (canDrain.Task.IsCompleted)
+            if (canDrainAny.Task.IsCompleted)
                 return true;
 
             using (var cts = new CancellationTokenSource())
             {
-                var delay = Task.Delay(timeout, cts.Token);
+                var waitTimeout = Task.Delay(timeout, cts.Token);
 
-                var result = await Task.WhenAny(canDrain.Task, delay).ConfigureAwait(false);
-                if (result == delay)
+                var result = await Task.WhenAny(canDrainAny.Task, waitTimeout).ConfigureAwait(false);
+                if (result == waitTimeout)
                     return false;
 
                 cts.Cancel();
                 return true;
             }
         }
+        
+        /// <summary>
+        /// <para>Asynchronously waits until something is available to <see cref="Drain"/> or the provided <paramref name="timeout"/> expires.</para>
+        /// <para>If there are less than batch items to drain, waits <paramref name="delay"/> delay.</para>
+        /// </summary>
+        /// <returns><c>true</c> if there is something to drain, <c>false</c> otherwise.</returns>
+        public async Task<bool> TryWaitForNewItemsBatchAsync(TimeSpan delay, TimeSpan timeout)
+        {
+            if (canDrainBatch.Task.IsCompleted)
+                return true;
+
+            using (var cts = new CancellationTokenSource())
+            {
+                var waitTimeout = Task.Delay(timeout, cts.Token);
+                var waitDelay = Task.Delay(delay, cts.Token);
+
+                await Task.WhenAny(canDrainBatch.Task, waitDelay, waitTimeout).ConfigureAwait(false);
```

Comment:
What's the point of using await of the minimum of delay/timeout?
We also do not need two tasks here, because we can say right from the start which will end first.